70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+
+		auto updatePresentationTimestamp = [&]()
+		{
+			m_currentImageAcquire = m_surface->acquireNextImage();
+
+			oracle.reportEndFrameRecord();
+			const auto timestamp = oracle.getNextPresentationTimeStamp();
+			oracle.reportBeginFrameRecord();
+
+			return timestamp;
+		};
+
+		const auto nextPresentationTimestamp = updatePresentationTimestamp();
+
+		if (!m_currentImageAcquire)
+			return;
+
+		auto* const cmdbuf = m_cmdBufs.data()[resourceIx].get();
+		cmdbuf->reset(IGPUCommandBuffer::RESET_FLAGS::RELEASE_RESOURCES_BIT);
+		cmdbuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+		cmdbuf->beginDebugMarker("Frame Debug FLIP sim begin");
+		{
+			camera.beginInputProcessing(nextPresentationTimestamp);
+			mouse.consumeEvents([&](const IMouseEventChannel::range_t& events) -> void { camera.mouseProcess(events); mouseProcess(events); }, m_logger.get());
+			keyboard.consumeEvents([&](const IKeyboardEventChannel::range_t& events) -> void { camera.keyboardProcess(events); }, m_logger.get());
+			camera.endInputProcessing(nextPresentationTimestamp);
+		}
+
+		SMVPParams camData;
+		SBufferRange<IGPUBuffer> camDataRange;
+		{
+			const auto viewMatrix = camera.getViewMatrix();
+			const auto projectionMatrix = camera.getProjectionMatrix();
+			const auto viewProjectionMatrix = camera.getConcatenatedMatrix();
+
+			core::matrix3x4SIMD modelMatrix;
+			modelMatrix.setTranslation(nbl::core::vectorSIMDf(0, 0, 0, 0));
+			modelMatrix.setRotation(quaternion(0, 0, 0));
+
+			core::matrix3x4SIMD modelViewMatrix = core::concatenateBFollowedByA(viewMatrix, modelMatrix);
+			core::matrix4SIMD modelViewProjectionMatrix = core::concatenateBFollowedByA(viewProjectionMatrix, modelMatrix);
+
+			auto modelMat = core::concatenateBFollowedByA(core::matrix4SIMD(), modelMatrix);
+
+			const core::vector3df camPos = camera.getPosition().getAsVector3df();
+
+			camPos.getAs4Values(camData.cameraPosition);
+			memcpy(camData.MVP, modelViewProjectionMatrix.pointer(), sizeof(camData.MVP));
+			memcpy(camData.M, modelMat.pointer(), sizeof(camData.M));
+			memcpy(camData.V, viewMatrix.pointer(), sizeof(camData.V));
+			memcpy(camData.P, projectionMatrix.pointer(), sizeof(camData.P));
+			{
+				camDataRange.buffer = cameraBuffer;
+				camDataRange.size = cameraBuffer->getSize();
+
+				cmdbuf->updateBuffer(camDataRange, &camData);
+			}
+		}
+
+		bool bCaptureTestInitParticles = false;
+		float32_t4 pressureSolverParams[2];
+		SBufferRange<IGPUBuffer> gridDataRange;
+		SBufferRange<IGPUBuffer> pParamsRange;
+		SBufferRange<IGPUBuffer> pressureParamsRange;
+		if (m_shouldInitParticles)
+		{
+			bCaptureTestInitParticles = true;
+
+			{
+				gridDataRange.size = gridDataBuffer->getSize();
+				gridDataRange.buffer = gridDataBuffer;
+			}
+			cmdbuf->updateBuffer(gridDataRange, &m_gridData);
+
+			{
+				pParamsRange.size = pParamsBuffer->getSize();
+				pParamsRange.buffer = pParamsBuffer;
+			}
+			cmdbuf->updateBuffer(pParamsRange, &m_pRenderParams);
```

Comment:
there's a `m_utils` in your Base class, these are `IUtilities` which have a method to create a buffer on DeviceLocal memory with some content already, see the other examples